### ISSUE TITLE
Strange behavior of ltac redefinitions w.r.t. `Import`

### ISSUE THREAD

**Issue by maximedenes** [coq/coq#10556]
I find the following behavior puzzling:

```coq
Module M.
Ltac foo := fail.
End M.

Module N.
Import M.
Ltac foo ::= exact I.
End N.

Goal True.
M.foo. (* succeeds *)
Qed.

Import M.

Goal True.
Fail M.foo.
```

@ppedrot is it a bug?

**Comment by maximedenes** [comment]
Incidentally, this bug shows up in Equations once #10480 is fixed...

**Comment by ppedrot** [comment]
This looks suspicious. Importing the defining module should not change the behaviour of the tactic, the expected semantics is the one of late binding.

**Comment by maximedenes** [comment]
@ppedrot please tell me that Ltac2 has no such late binding. Then we can probably forget about this issue and encourage people to migrate.

**Comment by ppedrot** [comment]
Potentially sorry about that, but Ltac2 does have this kind of feature. It's useful to write extensible code, and for technical reasons Ltac2 prevents mutable references to be declared at toplevel, so you need a rebinding mechanism.

**Comment by maximedenes** [comment]
So Ltac2 has semantically dubious features (late-binding) because of technical reasons? It reminds me of Ltac1 ^^

What are these technical limitations? How hard would it be to overcome them?

**Comment by ppedrot** [comment]
> So Ltac2 has semantically dubious features

Nah, it's not that semantically dubious. First, contrarily to Ltac1, only definitions marked as such can be rebound, and furthermore the overall mechanism is straightforward. I wouldn't be bothered too much about it personally.

> What are these technical limitations? How hard would it be to overcome them?

It's essentially the OCaml runtime, so, not solvable. There is data that cannot be marshalled, e.g. locations are desynchronized by it.

**Comment by maximedenes** [comment]
It seems Ltac2 has the same problem:

```coq
From Ltac2 Require Import Ltac2.

Module M.
Ltac2 mutable foo := fun () => ltac1:(fail).
End M.

Module N.
Ltac2 Set M.foo := fun () => ltac1:(exact I).
End N.

Import N.

Goal True.
M.foo ().
Qed.

Import M.

Goal True.
Fail M.foo ().
```

**Comment by maximedenes** [comment]
What should I do to fix this?

**Comment by ppedrot** [comment]
> It seems Ltac2 has the same problem

I am not surprised, it reuses a similar code.

> What should I do to fix this?

First, decide of the expected behaviour. I am not even sure what we want exactly.

**Comment by mattam82** [comment]
The feature we want is to have the latest redefinition performed. In the case of Equations there should only be one redefinition though, but e.g for the LibHypsRenaming library you need multiple redefinitions. I would guess Importing the original module where the mutable tactic is originally defined should not change the binding.

**Comment by SkySkimmer** [comment]
https://github.com/coq/coq/pull/18713 changes this for Ltac2
Not sure we want to touch Ltac1.
